gms/gossiper.cc

```diff
@@ -1500,12 +1526,35 @@ const endpoint_state& gossiper::get_endpoint_state(inet_address ep) const {
     return *it->second;
 }
 
-endpoint_state& gossiper::get_or_create_endpoint_state(inet_address ep) {
-    auto it = _endpoint_state_map.find(ep);
-    if (it == _endpoint_state_map.end()) {
-        it = _endpoint_state_map.emplace(ep, make_endpoint_state_ptr({})).first;
+endpoint_state& gossiper::insert_endpoint_state(const endpoint_id& node, endpoint_state_ptr eps) {
+    endpoint_state_ptr prev;
+    auto [it, inserted] = _endpoint_state_map_by_host_id.try_emplace(node.host_id, eps);
+    if (!inserted) {
+        prev = std::exchange(it->second, eps);
+    } else {
+        logger.debug("Inserted endpoint_state for node {}, status={}", node, get_gossip_status(*eps));
     }
-    return const_cast<endpoint_state&>(*it->second);
+    try {
+        _endpoint_state_map[node.addr] = eps;
+    } catch (...) {
+        if (prev) {
+            it->second = std::move(prev);
+        }
```

Comment:
Should we remove from `_endpoint_state_map_by_host_id` if `inserted`?

Reply:
yes, thanks.